File: src/api/data-sync.ts
import { Channel } from '@tauri-apps/api/core'
import { invokeCommand } from '@/api/base'
import type { SyncConfig, SyncPreview, SyncProgress } from '@/types/data-sync'

/**
 * 预览同步计划
 *
 * 展示每张表的行数、列信息、主键等预览信息，
 * 供用户在执行同步前确认同步范围。
 *
 * @param config 同步配置
 * @returns 每张表的预览信息列表
 */
export function syncDataPreview(config: SyncConfig): Promise<SyncPreview[]> {
  return invokeCommand('sync_data_preview', {
    sourceConnectionId: config.sourceConnectionId,
    sourceDatabase: config.sourceDatabase,
    targetConnectionId: config.targetConnectionId,
    targetDatabase: config.targetDatabase,
    tables: config.tables,
    syncMode: config.syncMode,
  }, { source: 'DATABASE' })
}

/**
 * 执行数据同步
 *
 * 通过 Tauri Channel 实时推送同步进度到前端。
 *
 * @param config 同步配置
 * @param onProgress 进度回调
 * @returns 执行结果摘要
 */
export function syncDataExecute(
  config: SyncConfig,
  onProgress: (progress: SyncProgress) => void,
): Promise<string> {
  const channel = new Channel<SyncProgress>()
  channel.onmessage = onProgress
  return invokeCommand('sync_data_execute', {
    config,
    onProgress: channel,
  }, { source: 'DATABASE' })
}

File: src-tauri/src/commands/data_sync.rs
use tauri::{command, ipc::Channel, AppHandle, Manager};

use crate::commands::db::DbEngineState;
use crate::models::scheduler::{SyncConfig, SyncPreview, SyncProgress};
use crate::services::data_sync;
use crate::utils::error::AppError;

/// 预览同步计划：展示每张表的行数、列信息、主键
#[command]
pub async fn sync_data_preview(
    app: AppHandle,
    source_connection_id: String,
    source_database: String,
    target_connection_id: String,
    target_database: String,
    tables: Vec<String>,
    sync_mode: String,
) -> Result<Vec<SyncPreview>, AppError> {
    let engine = app.state::<DbEngineState>().inner().clone();

    let config = SyncConfig {
        source_connection_id,
        source_database,
        target_connection_id,
        target_database,
        tables,
        sync_mode,
        page_size: None,
    };

    data_sync::sync_preview(engine, &config).await
}

/// 执行数据同步，通过 Channel 推送进度
#[command]
pub async fn sync_data_execute(
    app: AppHandle,
    config: SyncConfig,
    on_progress: Channel<SyncProgress>,
) -> Result<String, AppError> {
    let engine = app.state::<DbEngineState>().inner().clone();

    let progress_callback = move |progress: SyncProgress| {
        let _ = on_progress.send(progress);
    };

    data_sync::sync_tables(engine, &config, progress_callback).await
}
